### ISSUE TITLE
[V5] Error with redux-loop

### ISSUE THREAD

**Issue by ercpereda** [rt2zz/redux-persist#501]
redux-loop combineReducers no return a plain object

**Comment by lsunsi** [comment]
Yeah I was so happy about the update and then I got less happy. (:

**Comment by rt2zz** [comment]
I have not used redux-loop, does it add functions to each sub reducer?

Also what is the specific error you are getting? the default serialization is JSON.serialize so it should just omit any methods and does not necessarily need to result in error.

**Comment by ercpereda** [comment]
@rt2zz The problem is with this check https://github.com/rt2zz/redux-persist/blob/master/src/persistReducer.js#L59

Using `redux-loop` the `defaultState` is an array, with the actual `defaultState` object in the first element and some `Effects` in the second one.

Let me know if you need some help to make a `redux-loop` integration.

**Comment by rt2zz** [comment]
ah I see. we store state in _persist as a peer to the other keys, which is a big reason we need it to be a plain object. It seems like redux-loop perhaps only temporarily changes the state shape? Im not 100% sure what integration would look like, but I would be open to reviewing a PR. 

**Note** we are unlikely to take on any changes that will impact performance or have a large scope.

**Comment by ercpereda** [comment]
@rt2zz thanks for taking care of this. I understand that changes that impact in performance would be a problem.
Just like a note, `redux-loop` works excellent with `redux-persist` in the previous version.

**Comment by rt2zz** [comment]
ya, it may be that v4 is better fit for this case since it doesnt store any persist state. Will require more investigation to know for certain what the options are 👍

**Comment by lsunsi** [comment]
@rt2zz @ercpereda About not knowing what the integration would look like, what if redux-persist took the state getter/setter argument in some way? Instead of assuming it'll be an object and getting and setting to it, it could take the functions that take the state as parameters or something.

```javascript
// defaults that allow current behaviour
{ get: state => state._persist,
, set: (state, _persist) => ({...state, ._persist})
}

//
{ get: state => state[1]._persist,
  set: (state, _persist) => [{...state[0], _persist}, state[1]]
} 
```

This would allow for any kind of customised state shape to be compatible with persist, not only about redux-loop.

**Comment by johhansantana** [comment]
was this ever fixed? Any recommendation on how to use `redux-loop` with `redux-persist`? I have currently v6 installed. Would downgrading to v4 work?

**Comment by bdwain** [comment]
I'm not super familiar with redux-persist, so i can't say for certain, but i think the suggestion made by @lsunsi would be helpful. 

redux-loop expects your top level reducer to return a pair (2 item array) of `[state, Effect]` and it takes that pair, updates the state, and executes the effect. 

It sounds like redux-persist had an expectation that the top level state would be an object? That would cause problems potentially even outside of redux-loop. But if you exposed a way to pass a getter for the new state from the return value, it could allow you to tell redux-persist how to read the state from the return value. redux-loop exposes a method called `getModel` that gives you the state from the return value, so that would allow someone using both libraries to tell redux-persist how to read the state.

I know it's a little odd that the reducer would not return state, but it's a core part of the API for redux-loop so it's not something that can change. i'm not sure how useful this would be outside of integrating with redux-loop, but it would at least solve the issue without having to depend on any implementation details, and would be useful for any other library that modified the return shape.

**Comment by johhansantana** [comment]
hey @bdwain thanks for commenting. I just tried using the `getModel` and it seems to return the right updated state but now the `Cmd.action` gets executed but it seems like it doesn't recognize it as a dispatch or something because it's not reaching the reducer `switch` area.

This is in my reducer

```javascript
const stateUpdated = getModel(
    loop(
        newState,
        Cmd.action(
            (() => {
                return {
                    type: SOME_TYPE,
                    payload: somePayload
                };
            })()
        )
    )
);
return stateUpdated;
```
